src/CommunityToolkit/Xamarin.CommunityToolkit/ObjectModel/Internals/BaseCommand.shared.cs

```diff
@@ -83,7 +87,10 @@ protected int ExecutionCount
 		public void RaiseCanExecuteChanged()
 		{
 			// Automatically marshall to the Main Thread to adhere to the way that Xamarin.Forms automatically marshalls binding events to Main Thread
-			Device.BeginInvokeOnMainThread(() => weakEventManager.RaiseEvent(this, EventArgs.Empty, nameof(CanExecuteChanged)));
+			if (synchronizationContext != null && synchronizationContext != SynchronizationContext.Current)
```

Comment:
now it doesn't marshall to the Main Thread (cuz synchronizationContext can be not UI thread's context)

Reply:
well I think the Post method of SyncContext does run the task in the android message loop (aka UI thread).

https://github.com/xamarin/xamarin-android/blob/aad2c29e30e362b19bf526eb14075d8325a1f2d8/src/Mono.Android/Android.App/SyncContext.cs#L30

		public override void Post (SendOrPostCallback d, object state)
		{
			var looper = Application.Context?.MainLooper;
			if (!EnsureLooper (looper, d))
				return;
			using (var h = new Handler (looper)) {
				h.Post (() => d (state));
			}
		}

Reply:
Hey Gang! I'm going to change this to use the platform-specific code for both checking if the current thread is the Main Thread and invoking Main Thread.

It'll be inspired by Xamarin.Essentials.MainThread: https://github.com/xamarin/Essentials/tree/main/Xamarin.Essentials/MainThread

Reply:
The bad part of it's now we have 3 implementations of the same thing, on Forms, essentials and here... I would say to mock the Device platform services as @brminnick said on issue and @AndreiMisiukevich here...

Reply:
Well to be honest the point raised by @WebDucer is valid: a command belongs to the view model layer and then should be UI independent. The synchronization context is the correct way to implement main thread Marshalling in the net standard fashion. Since Post method make sure the task is executed on the main thread, this code is valid and does not need any modification (in fact this code is taken from prism implementation of delegate command...). Essentials and Forms are UI frameworks so it's only natural they are using directly Xamarin.Forms Device helpers.

Reply:
@pictos I agree that having 3 implementations is not idea, but I see it as a short-term stop-gap.

.NET MAUI will include Xamarin.Essentials, so once we migrate Xamarin.CommunityToolkit from Xamarin.Forms v5.0 to .NET MAUI, we can replace this duplicate implementation with Xamarin.Essentials.MainThread's implementation.

Reply:
I'm afraid the net standard implementation of the MainThread of essentials will not work: 

       static bool PlatformIsMainThread =>
            throw ExceptionUtils.NotSupportedOrImplementedException

So without SynchronizationContext object, it will force people unit testing view models to mock essentials or forms.

Reply:
@brminnick if you think this is a valuable change, let's keep it then (:

Reply:
@roubachof yeah, for forms people need to wrap the APIs inside interfaces if they want to test it